docs/design/tanka.md

```diff
@@ -0,0 +1,230 @@
+which may be composed into applications or modules which in turn may be applied
+to multiple environments.
+
+We believe such a concept overcomplicates the immediate goal of reducing
+duplication while allowing edge-cases, because it handles these on a higher
+conceptual level, instead of reusing the native capabilities of jsonnet.
+
+Code-reuse and composability is
+adequately provided by the native `import` feature of `jsonnet`. Sharing code
+beyond application boundaries is already enabled by `jsonnet-bundler`.
+
+While it is possible to mimic environments with native `jsonnet`, it falls short
+when it comes to actually applying it to the correct cluster. The raw `json`
+being returned by the compiler needs reconciling and it must be made sure it is
+applied to the correct cluster.
+
+This leaves us effectively with **`jsonnet`** and **[Environments](#environments)**
+
+## Code Reuse
+By using [`jsonnet`](https://jsonnet.org) as the underlying data templating
+language, tanka supports dynamic reusing of code, just like real programming languages do:
+
+### Imports
+`jsonnet` is able to
+[import](https://jsonnet.org/learning/tutorial.html#imports) other jsonnet
+snippets into the current scope, which in turn allows to refactor commonly used code
+into shared libraries.
+
+### Bundle
+Once a library becomes general enough to be used beyond project or
+even domain boundaries, it is a common practice in programming languages to have
+shared dependencies.
+
+In the `jsonnet` world, this is provided by
+[`jsonnet-bundler`](https://github.com/jsonnet-bundler/jsonnet-bundler), which maintains a
+`vendor` folder (the bundle) with all required libraries ready to be imported,
+much like older versions of `go` used to (>1.11) do.  
+This procedure integrates smoothly with tanka, because `vendor` is on the [`JPATH`](#jpath).
+
+### `JPATH`
+To enable a predictable developer experience, tanka uses clear rules to define
+how importing works.
+
+Imports are relative to the `JPATH`. The earlier a directory appears in the
+`JPATH`, the higher it's precedence is.
+
+To set it up, tanka makes use of the following directoies:
+
+| Name             | Identifier         | Description                                                                                                                           |
+|------------------|--------------------|---------------------------------------------------------------------------------------------------------------------------------------|
+| `rootDir`        | `jsonnetfile.json` | Every file in the tree of this folder is considered part of the project. Much like `git` has the one directory with the `.git` folder |
+| `rootDir/vendor` |                    | Populated with shared dependencies by `jsonnet-bundler`                                                                               |
+| `baseDir/units`  |                    | Code that directly outputs Kubernetes objects (mixins, etc.)
```

Comment:
Where does this come from?  Do we use this?

Reply:
We use "environments" IIRC.  Might want to mention you can do this, and that this is backwards compatible with ksonnet.

Reply:
OIC mentioned below.

Reply:
Was a personal wish from @malcolmholmes, because he did not like to put "components" into the lib folder, because that folder was called lib. 

So now there is another folder that is basically lib as well, but just not called lib so you can put something that is not a library in there if you want

Reply:
If you don't want to, just dump it into lib

Reply:
IMO we should drop this and make it clear you should put custom components in lib.  No point in having two dir that do the same thing.

And make it clear jb owns vendor/ but the user owns lib/

Reply:
/cc @malcolmholmes

Reply:
I will state that I don't particularly like putting descriptions of how to install my apps into `/lib`. I'm producing "components" of my systems, not "libraries". To me a "library" is a tool that I use to do stuff, not a description of how to install a component of my system.

I totally get that the term "component" is polluted in this space, hence @sh0rez's suggestion of 'unit' instead.

Reply:
@malcolmholmes IMO the point of this tool is to try and not introduce further abstractions. I'd say "unit" implies too much here.  This is not to say we can't introduce it in the future, but for now, lets stay minimal.

Reply:
:+1: